# Problem Set 01: Introduction and Getting Started - comments/solutions

# Programming for Economists Autumn 2026: Class 3
**Matias BF Hall**


This problem set uses material from the introduction lecture and *Introduction to Python*, chapters 1 and 2 on DataCamp. You will work with variables, arithmetic, lists, indexing and slicing. Do not use loops, functions, or NumPy.

Use descriptive variable names, add short comments where they help, and label the values you print. The answer sections contain complete code solutions. Run your own code before opening them.

**Table of contents**<a id='toc0_'></a>    
1. [Before you start](#toc1_)    
2. [Links to guides](#toc2_)    
3. [Predict what the code prints](#toc3_)    
4. [Calculate monthly transport costs](#toc4_)    
5. [Use indexes and slices](#toc5_)    
6. [Correct and summarize a spending series](#toc6_)    
7. [Find and fix the errors](#toc7_)    

<!-- vscode-jupyter-toc-config
	numbering=true
	anchor=true
	flat=true
	minLevel=2
	maxLevel=6
	/vscode-jupyter-toc-config -->
<!-- THIS CELL WILL BE REPLACED ON TOC UPDATE. DO NOT WRITE YOUR TEXT IN THIS CELL -->

## 1. <a id='toc1_'></a>[Before you start](#toc0_)

Do these three things first; everything below assumes they are done.

1. **Install Anaconda and VS Code.** Anaconda gives you Python and the scientific packages, VS Code is the editor you will write in. Section 0 of `01_Introduction/introduction.ipynb` walks through downloading the course material Here is the [Installation Guide](https://sites.google.com/view/numeconcph-progecon/guides/installation).
2. **Create your own notebook and run something in it.** Make a new Jupyter notebook in VS Code, copy/type the four lines below into a cell and run it. If `3` appears underneath, your installation works.
Easiest way to create a jupyter notebook in VSCode is by pressing (Ctrl/Cmd+Shift+P) and running the command "Create: New Jupyter Notebook"

In [1]:
a = 1
b = 2
c = a + b
print(c)

3


3. **Set up Restart-and-Run-All.** In the VS Code keyboard shortcuts, bind *Jupyter: Restart Kernel and Run All Cells*. You will finish every problem set in this course with it. It is the only reliable way to see whether a notebook works from a clean start rather than from whatever order you happened to click.

## 2. <a id='toc2_'></a>[Links to guides](#toc0_)
Follow the guide for [VS Code](https://sites.google.com/view/numeconcph-progecon/guides/vscode)  
Use this guide for looking up the formatting of **Markdown**: [Markdown Guide](https://www.markdownguide.org/basic-syntax/)  
If using Git find the guide here: [Git Guide](https://sites.google.com/view/numeconcph-progecon/guides/git)

## 3. <a id='toc3_'></a>[Predict what the code prints](#toc0_)

Two short pieces of code. For each one, write down what you expect it to print **before** running anything — on paper or in the Markdown cell underneath. Then run it and compare.

A wrong prediction is useful; it tells you exactly which idea to go back to. A right prediction you did not think about tells you nothing.

In [2]:
# Solution with explanation

x = [1, 2, 3]
# creates a list x with three elements
y = x
# assigns the list x to y, y is a reference to the same place in memory as x
y[-1] = 4
# changes the last element of y to 4, which also changes the last element of x, since they refer to the same place in memory
print(x)
# [1, 2, 4]

[1, 2, 4]


In [3]:
# Solution with explanation

x = [0 ,1 ,2 ,3 ,4 ,5]
#defines a list x with six elements

print (x [:2])
# prints a slice of x - the first two elements in x
# [0,1]
print (x [2:])
# prints a slice of x - all elements from index 2 and onwards
# [2,3,4,5]

[0, 1]
[2, 3, 4, 5]


**My predictions:**

**Where I was wrong, and why:**

<details>
<summary><strong>Answer</strong></summary>

The first prints `[1, 2, 4]`. Most people predict `[1, 2, 3]` the first time. `y = x` does not create a second list — it gives the existing list a second name, so a change made through `y` is visible through `x`. Problem Set 02 covers what to write when you genuinely want an independent copy.

The second prints `[0, 1]` and then `[2, 3, 4, 5]`. A slice runs up to but not including its second position, and an empty side of the colon means "from the start" or "to the end". Notice that `x[:2]` and `x[2:]` split the list cleanly: nothing is counted twice and nothing is lost. You will use exactly that in Task 4.

</details>

## 4. <a id='toc4_'></a>[Calculate monthly transport costs](#toc0_)

A student travels to campus on 18 days during a month and pays for two trips on each travel day. A single trip costs 24 DKK. In the following month, the fare rises by 12.5 percent and the student travels on 16 days.

1. Calculate the original monthly cost.
2. Calculate the new fare without overwriting the original fare.
3. Calculate the new monthly cost using 16 travel days.
4. Calculate the change in monthly cost.
5. Calculate what the monthly cost would have been if the fare had increased but the student had still travelled on 18 days. Store it as `price_only_cost`.
6. Print the results with clear labels.

In a short Markdown note, explain why `price_only_cost` differs from the new monthly cost.

In [4]:
single_trip_fare = 24.0
travel_days = 18
trips_per_day = 2
fare_increase_rate = 0.125
new_travel_days = 16

In [5]:
# Solution with explanation

# trip cost before the fare increase
trip_cost_before = single_trip_fare * travel_days * trips_per_day

# new trip cost after the fare increase
new_single_trip_fare = single_trip_fare * (1 + fare_increase_rate)

# new trip cost after the fare increase and reduced travel days
new_trip_cost = new_single_trip_fare * new_travel_days * trips_per_day

# new trip cost after the fare increase and reduced travel days
price_only_cost = new_single_trip_fare * travel_days * trips_per_day

# print results
# we use f-strings to format the output and include the variable values in the printed message
# with f-strings we can include expressions inside curly braces {} and they will be evaluated and included in the string

print('Original monthly travel:')
print(f'When traveling {travel_days} days, {trips_per_day} times a day with a fare price of {single_trip_fare}, the total cost is: {trip_cost_before}\n')

print('New monthly travel:')
print(f'After a fare increase of {fare_increase_rate*100}%, the new fare is: {new_single_trip_fare}')
print(f'When traveling {new_travel_days} days, {trips_per_day} times a day with the new fare price of {new_single_trip_fare}, the total cost is: {new_trip_cost}\n')

print('Price only travel:')
print(f'If we only consider the fare increase and not the reduced travel days, the total cost would be: {price_only_cost}')

# I intepret the results as follows: The agent substitutes away from spending money on travel when the fare increases by reducing the amount of days they travel.
# the agent ends up spending the same amount on travel as before the fare increase, so perhaps the agent has a budget for travel, that they are willing to spend
# always to avoid having to bike/walk as much as possible.

Original monthly travel:
When traveling 18 days, 2 times a day with a fare price of 24.0, the total cost is: 864.0

New monthly travel:
After a fare increase of 12.5%, the new fare is: 27.0
When traveling 16 days, 2 times a day with the new fare price of 27.0, the total cost is: 864.0

Price only travel:
If we only consider the fare increase and not the reduced travel days, the total cost would be: 972.0


**Explanation:**

<details>
<summary><strong>Answer</strong></summary>

```python
single_trip_fare = 24.0
travel_days = 18
trips_per_day = 2
fare_increase_rate = 0.125
new_travel_days = 16

original_monthly_cost = travel_days * trips_per_day * single_trip_fare
new_single_trip_fare = single_trip_fare * (1 + fare_increase_rate)
new_monthly_cost = new_travel_days * trips_per_day * new_single_trip_fare
change_in_monthly_cost = new_monthly_cost - original_monthly_cost
price_only_cost = travel_days * trips_per_day * new_single_trip_fare

print("Original monthly cost:", original_monthly_cost, "DKK")
print("New fare per trip:", new_single_trip_fare, "DKK")
print("New monthly cost:", new_monthly_cost, "DKK")
print("Change in monthly cost:", change_in_monthly_cost, "DKK")
print("Cost with only the fare change:", price_only_cost, "DKK")
```

`price_only_cost` keeps 18 travel days, while the new monthly cost uses 16. The higher fare therefore applies to fewer trips.

Notice that the change in monthly cost comes out as exactly `0.0`. Your code is fine. The fare rise adds 108 DKK at 18 travel days (864 to 972), and cutting the month to 16 travel days takes the same 108 DKK away again (972 back to 864). Separating a change into the part caused by the price and the part caused by the quantity is something you will do again and again in economics; here the two happen to cancel exactly.

</details>

## 5. <a id='toc5_'></a>[Use indexes and slices](#toc0_)

The two lists below contain illustrative unemployment rates for five countries. The country and its rate have the same position in the two lists.

1. Correct the four commented examples in the empty code cell. Each correction needs only one or two lines.
2. Run the corrected examples and check the output against the two lists.
3. Change `selected_position` from `2` to `0` and then to `-1`. Run the same print statement after each change.
4. Change the slice so that it selects the first three countries instead of the middle three. Then change it back.

In one sentence, explain why the same position must be used for both lists.

In [6]:
# Solution with explanation

country_names = ["Denmark", "Sweden", "Germany", "Netherlands", "France"]
unemployment_rates = [6.1, 8.4, 6.4, 3.8, 7.5]

# a)
# Intended: print the first country and its rate
# print(country_names[1], unemployment_rates[0])
# correction:
print(country_names[0], unemployment_rates[0])
# explanation: indexes run from 0 to n-1

# b)
# Intended: print the last country and its rate
# print(country_names[5], unemployment_rates[5])
# correction:
print(country_names[-1], unemployment_rates[-1])
# explanation: negative indexes count from the end of the list, so -1 is the last element

# c)
# Intended: print Sweden, Germany, and Netherlands
# print(country_names[1:3])
# correction:
print(country_names[1:4])
# explanation: when indexing a list, the first index is inclusive and the second is exclusive

# d)
# Intended: use one position to print Germany and its rate
# selected_position = "2"
# print(country_names[selected_position], unemployment_rates[selected_position])
# correction:
selected_position = 2
print(country_names[selected_position], unemployment_rates[selected_position])
# explanation: the index must be an integer, not a string

Denmark 6.1
France 7.5
['Sweden', 'Germany', 'Netherlands']
Germany 6.4


**Explanation:**

<details>
<summary><strong>Answer</strong></summary>

```python
country_names = ["Denmark", "Sweden", "Germany", "Netherlands", "France"]
unemployment_rates = [6.1, 8.4, 6.4, 3.8, 7.5]

print("First:", country_names[0], unemployment_rates[0])
print("Last:", country_names[-1], unemployment_rates[-1])
print("Middle three:", country_names[1:4])

selected_position = 2
print("Selected:", country_names[selected_position], unemployment_rates[selected_position])

# Change selected_position and the slice above, then rerun the same lines.
```

The two lists are parallel, so a country and its rate must be selected with the same position.

</details>

## 6. <a id='toc6_'></a>[Correct and summarize a spending series](#toc0_)

The lists below record monthly transport spending from January to June. The April value contains an extra zero.

1. Print the month and spending value for April using list indexes.
2. Calculate the average spending before correcting the error. Use `sum()` and `len()`.
3. Correct the April value in place. Do not retype the list.
4. Calculate the corrected average.
5. Use slices to calculate the average for the first three months and the last three months separately.
6. Calculate the percentage change from January to June.
7. Print the two slices and all calculated values with labels.

Add one sentence explaining why checking individual observations can matter before calculating a summary statistic.

In [7]:
# Solution with explanation

# keep the original lists in the cell, to avoid having to re-run the previous cells if you want to re-run this cell
months = ["January", "February", "March", "April", "May", "June"]
transport_spending = [672.0, 720.0, 768.0, 7440.0, 816.0, 720.0]
print('Indexing/slicing method with lists:')

# 1.
print(f'Transport spending in {months[3]} before correction: {transport_spending[3]}')

# 2.
average_transport_spending = sum(transport_spending) / len(transport_spending)
# sum() adds all values, while len() returns the number of values in the list
print(f'Average transport spending before correcting the error: {average_transport_spending}')

# 3.
transport_spending[3] = 744.0
print(f'Transport spending in {months[3]} after correction: {transport_spending[3]}')

# 4.
average_transport_spending_corrected = sum(transport_spending) / len(transport_spending)
print(f'Average transport spending after correcting the error: {average_transport_spending_corrected}')

# 5.
average_first_three = sum(transport_spending[:3]) / len(transport_spending[:3])
average_last_three = sum(transport_spending[3:]) / len(transport_spending[3:])
print(f'Average transport spending for the first three months: {average_first_three}')
print(f'Average transport spending for the last three months: {average_last_three}')

# 6.
pct_change = transport_spending[-1]/transport_spending[0]-1
print(f'Percentage change in transport spending from January to June: {pct_change*100:.2f}%') # .2f formats the number to a float with 2 decimals
# 7. is done under each step

# EXTRA
# to easilier select the correct month a dictionary can be used instead

transport_dict = dict(zip(months, transport_spending)) # this creates a dictionary where the keys are the months and the values are the transport spending for that month
# the zip function takes two lists and returns an iterator of tuples, where the first element of each tuple is from the first list and the second element is from the second list

months = ["January", "February", "March", "April", "May", "June"]
transport_spending = [672.0, 720.0, 768.0, 7440.0, 816.0, 720.0]

# now we can easily access the transport spending for a specific month by using the month name as the key

print('\n')
print('Dictionary method:')
# e.g. 1. now becomes:
print(f'Transport spending in April: {transport_dict["April"]}')
# 2. now becomes:
average_transport_spending_dict = sum(transport_dict.values()) / len(transport_dict)
print(f'Average transport spending: {average_transport_spending_dict}')
# 3. now becomes:
transport_dict["April"] = 744.0
print(f'Transport spending in April after correction: {transport_dict["April"]}')
# 4. now becomes:
average_transport_spending_corrected_dict = sum(transport_dict.values()) / len(transport_dict)
print(f'Average transport spending after correction: {average_transport_spending_corrected_dict}')
# 5. now becomes:
average_first_three_dict = sum(list(transport_dict.values())[:3]) / len(list(transport_dict.values())[:3])
average_last_three_dict = sum(list(transport_dict.values())[3:]) / len(list(transport_dict.values())[3:])
print(f'Average transport spending for the first three months: {average_first_three_dict}')
print(f'Average transport spending for the last three months: {average_last_three_dict}')
# 6. now becomes:
pct_change_dict = transport_dict["June"]/transport_dict["January"]-1
print(f'Percentage change in transport spending from January to June: {pct_change_dict*100:.2f}%')

# some things are easier with lists like indexing and slicing, while others are easier with dictionaries like accessing values by key.

Indexing/slicing method with lists:
Transport spending in April before correction: 7440.0
Average transport spending before correcting the error: 1856.0
Transport spending in April after correction: 744.0
Average transport spending after correcting the error: 740.0
Average transport spending for the first three months: 720.0
Average transport spending for the last three months: 760.0
Percentage change in transport spending from January to June: 7.14%


Dictionary method:
Transport spending in April: 744.0
Average transport spending: 740.0
Transport spending in April after correction: 744.0
Average transport spending after correction: 740.0
Average transport spending for the first three months: 720.0
Average transport spending for the last three months: 760.0
Percentage change in transport spending from January to June: 7.14%


**Explanation:**

<details>
<summary><strong>Answer</strong></summary>

```python
months = ["January", "February", "March", "April", "May", "June"]
transport_spending = [672.0, 720.0, 768.0, 7440.0, 816.0, 720.0]

print("Selected month:", months[3])
print("Entered spending:", transport_spending[3], "DKK")

average_before_correction = sum(transport_spending) / len(transport_spending)

transport_spending[3] = 744.0
corrected_average = sum(transport_spending) / len(transport_spending)

first_three_months = transport_spending[:3]
last_three_months = transport_spending[3:]
first_three_average = sum(first_three_months) / len(first_three_months)
last_three_average = sum(last_three_months) / len(last_three_months)

percentage_change = (
    (transport_spending[-1] - transport_spending[0])
    / transport_spending[0]
    * 100
)

print("Average before correction:", average_before_correction, "DKK")
print("Corrected spending:", transport_spending)
print("Corrected average:", corrected_average, "DKK")
print("First three months:", first_three_months)
print("Last three months:", last_three_months)
print("First-three-month average:", first_three_average, "DKK")
print("Last-three-month average:", last_three_average, "DKK")
print("Change from January to June:", percentage_change, "percent")
```

An incorrect observation can strongly distort every summary that uses it, even when the Python calculation itself is correct.

</details>

## 7. <a id='toc7_'></a>[Find and fix the errors](#toc0_)

The code below is meant to calculate gross earnings, the tax payment, net earnings and the net wage per hour, using the same figures as Task 5. It contains four separate problems.

Copy it into the empty code cell and run it. Only the first problem announces itself with an error message. Once you have fixed that one, the code runs all the way to the end without complaining — and still prints wrong numbers. Running without an error is not evidence that a calculation is right.

Fix one problem at a time, rerun after each fix, and check every printed value against the number you expect.

```python
hourly_wage = 145.0
hours_worked = "84"
tax_rate = 0.36

gross_earnings = hourly_wage + hours_worked
tax_payment = gross_earnings * tax_rate
net_earnings = tax_payment
net_hourly_wage = hourly_wage * 1 - tax_rate

print("Gross earnings:", gross_earnings, "DKK")
print("Tax payment:", tax_payment, "DKK")
print("Net earnings:", net_earnings, "DKK")
print("Net wage per hour:", net_hourly_wage, "DKK")
```

Write one sentence for each problem you corrected.

**Problems corrected:**

1. `TypeError`: `hours_worked` was a string instead of a number. Changed it to a float.
2. Multiplication v. addition: `gross_earnings` should be calculated with `*` instead of `+`.
3. Missing variable: `net_earnings` should be calculated as `gross_earnings - tax_payment` instead of just `tax_payment`.
4. Order of operations: `net_hourly_wage` should be calculated as `hourly_wage * (1 - tax_rate)` instead of `hourly_wage * 1 - tax_rate`.

The corrected code is found below.

In [8]:
hourly_wage = 145.0
hours_worked = 84.0
tax_rate = 0.36

gross_earnings = hourly_wage * hours_worked
tax_payment = gross_earnings * tax_rate
net_earnings = gross_earnings - tax_payment
net_hourly_wage = hourly_wage * (1 - tax_rate)

print("Gross earnings:", gross_earnings, "DKK")
print("Tax payment:", tax_payment, "DKK")
print("Net earnings:", net_earnings, "DKK")
print("Net wage per hour:", net_hourly_wage, "DKK")

Gross earnings: 12180.0 DKK
Tax payment: 4384.8 DKK
Net earnings: 7795.2 DKK
Net wage per hour: 92.8 DKK


<details>
<summary><strong>Answer</strong></summary>

```python
hourly_wage = 145.0
hours_worked = 84
tax_rate = 0.36

gross_earnings = hourly_wage * hours_worked
tax_payment = gross_earnings * tax_rate
net_earnings = gross_earnings - tax_payment
net_hourly_wage = hourly_wage * (1 - tax_rate)

print("Gross earnings:", gross_earnings, "DKK")
print("Tax payment:", tax_payment, "DKK")
print("Net earnings:", net_earnings, "DKK")
print("Net wage per hour:", net_hourly_wage, "DKK")
```

The four problems:

1. `hours_worked` held the text `"84"` instead of the number `84`, so the next line had a number and a piece of text to combine and stopped with a `TypeError`. This is the only one that produces an error message.
2. The wage and the hours were added instead of multiplied, which gives 229 DKK of gross earnings instead of 12,180 DKK.
3. `net_earnings` was set to the tax payment rather than to gross earnings minus the tax payment.
4. `hourly_wage * 1 - tax_rate` multiplies the wage by one and then subtracts 0.36, giving 144.64. The brackets in `hourly_wage * (1 - tax_rate)` are what turn it into the share of the wage left after tax, 92.80 DKK.

Corrected, the four printed values are `12180.0`, `4384.8`, `7795.2` and `92.8`.

</details>

Restart the kernel and run the notebook from top to bottom. Check that each print statement has a useful label and that changing an input changes every result that depends on it.